examples/game-of-life-challenge/src/game.rs

```diff
@@ -0,0 +1,1131 @@
+    /// Print the board as ASCII art.
+    /// Live cells are represented by '●' and dead cells by '·'.
+    fn to_pretty_string(&self) -> String {
+        let mut result = String::new();
+
+        // Add top border with column numbers for reference
+        result.push_str("   ");
+        for x in 0..self.size {
+            result.push_str(&format!("{:2}", x % 10));
+        }
+        result.push('\n');
+
+        for y in 0..self.size {
+            // Add row number for reference
+            result.push_str(&format!("{:2} ", y));
+
+            for x in 0..self.size {
+                result.push(' ');
+                if self.is_live(Position { x, y }) {
+                    result.push('●');
+                } else {
+                    result.push('·');
+                }
+            }
+            result.push('\n');
+        }
+
+        result
+    }
+
+    /// Print the board as compact ASCII art without numbers.
+    /// Live cells are represented by '●' and dead cells by '·'.
+    fn to_compact_string(&self) -> String {
+        let mut result = String::new();
+
+        for y in 0..self.size {
+            for x in 0..self.size {
+                if self.is_live(Position { x, y }) {
+                    result.push('●');
+                } else {
+                    result.push('·');
+                }
+            }
+            result.push('\n');
+        }
+
+        result
+    }
+
+    fn is_live(&self, position: Position) -> bool {
+        let Position { x, y } = position;
+        self.index.get(&x).is_some_and(|set| set.contains(&y))
+    }
+
+    fn check_conditions(
+        &self,
+        conditions: &[Condition],
+    ) -> Result<(), (usize, ConditionFailureReason)> {
+        for (index, condition) in conditions.iter().enumerate() {
+            condition.check(self).map_err(|reason| (index, reason))?;
+        }
+        Ok(())
+    }
+}
+
+impl Display for Board {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.index().fmt(formatter)
+    }
+}
+
+impl Board {
+    /// Creates a new board with the given size and player owners.
+    pub fn new(size: u16) -> Self {
+        Board {
+            size,
+            live_cells: BTreeSet::new(),
+        }
+    }
+
+    /// Returns all neighbors of cell `(x, y)`.
+    fn neighbors(&self, position: Position) -> impl Iterator<Item = (u16, u16)> {
+        let Position { x, y } = position;
+        iter::empty()
+            .chain((x > 0 && y > 0).then(|| (x - 1, y - 1)))
+            .chain((x > 0).then(|| (x - 1, y)))
+            .chain((x > 0 && y + 1 < self.size).then(|| (x - 1, y + 1)))
+            .chain((y > 0).then(|| (x, y - 1)))
+            .chain((y + 1 < self.size).then(|| (x, y + 1)))
+            .chain((x + 1 < self.size && y > 0).then(|| (x + 1, y - 1)))
+            .chain((x + 1 < self.size).then(|| (x + 1, y)))
+            .chain((x + 1 < self.size && y + 1 < self.size).then(|| (x + 1, y + 1)))
```

Comment:
Given the sparse representation, is there even a reason to have a size limit? Instead, we could limit the total number of live cells, and otherwise wrap the coordinates around.